README.md
# AnecdoteExtension

Extension adds welcome anecdotes when starting VS Code!

## Features

Read jokes from rzhunemogu.ru after launching VS Code and like your favorite.
You can reread the jokes you like by clicking on the widget in the status bar or by calling the command.

## Known Issues

Problems with updating page with favorite anecdotes while it webview is active or not in focused.

## Release Notes

### 0.0.1

Initial release of extension.

### 0.1.1

Minor features, fixes and improvements.

### 0.1.2

Add image icon to extension in marketplace.

### 0.1.3

Hotfix!

### 0.1.4

Actual Hotfix ¯\\\_(ツ)\_/¯

### 1.2.4

Breaking changes and new features

CHANGELOG.md
# Change Log

## [0.0.1]

- Initial release

## [0.1.1]

- Jokes.json file now contains no jokes after initial plugin installation
- New command to request a random joke - no more expensive reboots to see fresh joke!
- API calls rate to get next joke is limited to 0.3 s

## [0.1.2]

- Add image icon to extension in marketplace

## [0.1.3]

- Remove jokes.json from initial download in order not to remove favorite jokes on every update

## [0.1.4]

- Actual fix of accidental favorite jokes removal

## [1.1.4]

- Breaking changes: command `showJoke` renamed to `showFavorite`
- Now you can can call command `requestJoke` via presing button on status bar
- New image icon

src/extension.ts
import * as vscode from 'vscode';
import { join } from 'path';
import fetch from 'node-fetch';
import { TextDecoder } from 'util';
import * as fs from 'fs';

export async function activate(context: vscode.ExtensionContext) {
	showJoke(context);

	let currentPanel: vscode.WebviewPanel | undefined = undefined;

	context.subscriptions.push(vscode.commands.registerCommand('anecdotextension.requestJoke', () => showJoke(context)));

	context.subscriptions.push(
		vscode.commands.registerCommand('anecdotextension.showFavorite', () => {
			const columnToShowIn = vscode.window.activeTextEditor
				? vscode.window.activeTextEditor.viewColumn
				: undefined;

			if (currentPanel) {
				currentPanel.reveal(columnToShowIn);
			} else {
				currentPanel = vscode.window.createWebviewPanel(
					'anecdotextension', // Identifies the type of the webview; used internally
					'Anecdote Stats', // Title of the panel displayed to the user
					vscode.ViewColumn.One, // Editor column to show the new webview panel in
					{} // Webview options
				);

				const jokeArray = getJsonContent(join(context.extensionPath, 'jokes.json'));
				let spanArray = new Array<string>();

				for (let index = 0; index < jokeArray.length; index++) {
					spanArray.push(`<p>${jokeArray[index]}</p>`);
				}

				currentPanel.webview.html = getWebviewContent(spanArray.join(""));

				currentPanel.onDidDispose(
					() => {
						currentPanel = undefined;
					},
					null,
					context.subscriptions
				);
			}
		})
	);

	const statusBarItem1 = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
	statusBarItem1.text = 'Random Joke';
	statusBarItem1.command = 'anecdotextension.requestJoke';
	statusBarItem1.show();

	const statusBarItem2 = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
	statusBarItem2.text = 'Favorite Jokes';
	statusBarItem2.command = 'anecdotextension.showFavorite';
	statusBarItem2.show();
}

export function deactivate() { }

async function getJoke() {
	try {
		const response = await fetch('http://rzhunemogu.ru/RandJSON.aspx?CType=1', {
			method: 'GET',
			headers: {
				accept: 'application/json',
			},
		});

		if (!response.ok) {
			throw new Error('Error! status: ${response.status}');
		}

		const win1251decoder = new TextDecoder('windows-1251');
		var buf = await response.arrayBuffer();
		const textResponse = win1251decoder.decode(buf).replace('\"', '').replace('\"', '');

		return textResponse.substring(textResponse.indexOf('\"') + 1, textResponse.lastIndexOf('\"') - 1);
	} catch (error) {
		if (error instanceof Error) {
			console.log('error message: ', error.message);
		} else {
			console.log('unexpected error: ', error);
		}
		throw error;
	}
}

async function showJoke(context: vscode.ExtensionContext) {
	const joke = await getJoke();
	vscode.window.showInformationMessage(joke, 'Like', 'Next').then(arg => {
		if (arg === 'Next') {
			setTimeout(() => showJoke(context), 300);
		}
		else if (arg === 'Like') {
			addJsonContent(join(context.extensionPath, 'jokes.json'), joke);
		}
		else {
			throw new Error(`Unreachale: expected callback value were Like or Next, got ${arg}`);
		}
	}
	);
}

function getJsonContent(path: string) {
	checkJsonExists(path);
	const file = fs.readFileSync(path, 'utf8');
	return JSON.parse(file);
}

function addJsonContent(path: string, content: string) {
	checkJsonExists(path);
	const file = getJsonContent(path);
	file.push(content);
	fs.writeFileSync(path, JSON.stringify(file));
}

function checkJsonExists(path: string) {
	if (!fs.existsSync(path)) {
		fs.writeFileSync(path, '[]');
	}
}

function getWebviewContent(content: string) {
	return `<!DOCTYPE html>
  <html lang="en">
  <head>
	  <meta charset="UTF-8">
	  <meta name="viewport" content="width=device-width, initial-scale=1.0">
	  <title>Favorite Jokes</title>
  </head>
  <body>
	  	${content}
  </body>
  </html>`;
}

out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const path_1 = require("path");
const node_fetch_1 = require("node-fetch");
const util_1 = require("util");
const fs = require("fs");
async function activate(context) {
    showJoke(context);
    let currentPanel = undefined;
    context.subscriptions.push(vscode.commands.registerCommand('anecdotextension.requestJoke', () => showJoke(context)));
    context.subscriptions.push(vscode.commands.registerCommand('anecdotextension.showFavorite', () => {
        const columnToShowIn = vscode.window.activeTextEditor
            ? vscode.window.activeTextEditor.viewColumn
            : undefined;
        if (currentPanel) {
            currentPanel.reveal(columnToShowIn);
        }
        else {
            currentPanel = vscode.window.createWebviewPanel('anecdotextension', // Identifies the type of the webview; used internally
            'Anecdote Stats', // Title of the panel displayed to the user
            vscode.ViewColumn.One, // Editor column to show the new webview panel in
            {} // Webview options
            );
            const jokeArray = getJsonContent((0, path_1.join)(context.extensionPath, 'jokes.json'));
            let spanArray = new Array();
            for (let index = 0; index < jokeArray.length; index++) {
                spanArray.push(`<p>${jokeArray[index]}</p>`);
            }
            currentPanel.webview.html = getWebviewContent(spanArray.join(""));
            currentPanel.onDidDispose(() => {
                currentPanel = undefined;
            }, null, context.subscriptions);
        }
    }));
    const statusBarItem1 = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
    statusBarItem1.text = 'Random Joke';
    statusBarItem1.command = 'anecdotextension.requestJoke';
    statusBarItem1.show();
    const statusBarItem2 = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
    statusBarItem2.text = 'Favorite Jokes';
    statusBarItem2.command = 'anecdotextension.showFavorite';
    statusBarItem2.show();
}
exports.activate = activate;
function deactivate() { }
exports.deactivate = deactivate;
async function getJoke() {
    try {
        const response = await (0, node_fetch_1.default)('http://rzhunemogu.ru/RandJSON.aspx?CType=1', {
            method: 'GET',
            headers: {
                accept: 'application/json',
            },
        });
        if (!response.ok) {
            throw new Error('Error! status: ${response.status}');
        }
        const win1251decoder = new util_1.TextDecoder('windows-1251');
        var buf = await response.arrayBuffer();
        const textResponse = win1251decoder.decode(buf).replace('\"', '').replace('\"', '');
        return textResponse.substring(textResponse.indexOf('\"') + 1, textResponse.lastIndexOf('\"') - 1);
    }
    catch (error) {
        if (error instanceof Error) {
            console.log('error message: ', error.message);
        }
        else {
            console.log('unexpected error: ', error);
        }
        throw error;
    }
}
async function showJoke(context) {
    const joke = await getJoke();
    vscode.window.showInformationMessage(joke, 'Like', 'Next').then(arg => {
        if (arg === 'Next') {
            setTimeout(() => showJoke(context), 300);
        }
        else if (arg === 'Like') {
            addJsonContent((0, path_1.join)(context.extensionPath, 'jokes.json'), joke);
        }
        else {
            throw new Error(`Unreachale: expected callback value were Like or Next, got ${arg}`);
        }
    });
}
function getJsonContent(path) {
    checkJsonExists(path);
    const file = fs.readFileSync(path, 'utf8');
    return JSON.parse(file);
}
function addJsonContent(path, content) {
    checkJsonExists(path);
    const file = getJsonContent(path);
    file.push(content);
    fs.writeFileSync(path, JSON.stringify(file));
}
function checkJsonExists(path) {
    if (!fs.existsSync(path)) {
        fs.writeFileSync(path, '[]');
    }
}
function getWebviewContent(content) {
    return `<!DOCTYPE html>
  <html lang="en">
  <head>
	  <meta charset="UTF-8">
	  <meta name="viewport" content="width=device-width, initial-scale=1.0">
	  <title>Favorite Jokes</title>
  </head>
  <body>
	  	${content}
  </body>
  </html>`;
}
//# sourceMappingURL=extension.js.map